subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/junit/GenericJUnitTestEventAdapter.java

```diff
@@ -0,0 +1,108 @@
+import org.gradle.api.internal.tasks.testing.TestResultProcessor;
+import org.gradle.api.internal.tasks.testing.TestStartEvent;
+import org.gradle.api.tasks.testing.TestResult;
+import org.gradle.internal.time.Clock;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Adapt JUnit3/4 and JUnit Platform test event to Gradle test event.
+ *
+ * @param <T> the test identifier type
+ */
+public class GenericJUnitTestEventAdapter<T> {
+    private final TestResultProcessor resultProcessor;
+    private final Clock clock;
+    private final Object lock = new Object();
+    private final Map<T, TestDescriptorInternal> executing = new HashMap<T, TestDescriptorInternal>();
+    private final Set<T> assumptionFailed = new HashSet<T>();
+
+    public GenericJUnitTestEventAdapter(TestResultProcessor resultProcessor, Clock clock) {
+        assert resultProcessor instanceof org.gradle.internal.concurrent.ThreadSafe;
+        this.resultProcessor = resultProcessor;
+        this.clock = clock;
+    }
+
+    public void testStarted(T identifier, TestDescriptorInternal descriptor) {
+        synchronized (lock) {
+            TestDescriptorInternal oldTest = executing.put(identifier, descriptor);
+            assert oldTest == null : String.format("Unexpected start event for %s", identifier);
+        }
+        resultProcessor.started(descriptor, startEvent());
+    }
+
+    public void testFailure(T identifier, TestDescriptorInternal descriptor, Throwable exception) {
+        TestDescriptorInternal testInternal;
+        synchronized (lock) {
+            testInternal = executing.get(identifier);
+        }
+        boolean needEndEvent = false;
+        if (testInternal == null) {
+            // This can happen when, for example, a @BeforeClass or @AfterClass method fails
+            needEndEvent = true;
+            testInternal = descriptor;
+            resultProcessor.started(testInternal, startEvent());
+        }
+        resultProcessor.failure(testInternal.getId(), exception);
+        if (needEndEvent) {
+            resultProcessor.completed(testInternal.getId(), new TestCompleteEvent(clock.getCurrentTime()));
+        }
+    }
+
+    public void testAssumptionFailure(T identifier) {
+        synchronized (lock) {
+            assumptionFailed.add(identifier);
+        }
+    }
+
+    public void testIgnored(TestDescriptorInternal descriptor) {
+        resultProcessor.started(descriptor, startEvent());
+        resultProcessor.completed(descriptor.getId(), new TestCompleteEvent(clock.getCurrentTime(), TestResult.ResultType.SKIPPED));
+    }
+
+    public void testFinished(T identifier) {
+        long endTime = clock.getCurrentTime();
+        TestDescriptorInternal testInternal;
+        TestResult.ResultType resultType;
+        synchronized (lock) {
+            testInternal = executing.remove(identifier);
+            if (testInternal == null && executing.size() == 1) {
```

Comment:
⭕️ I know you just extracted this, but this is not going to work with parallel tests. Which is funny because all the synchronization blocks clearly indicate that we intend to support parallelism.

Reply:
Well, this looks out of this PR's scope. However, as far as I know, all test classes in a single worker process is executed sequentially, despite of the lock here. Do you mean parallel in each worker process?

Reply:
I meant parallel in a single worker, which can only happen with custom JUnit runners at the moment. It'll be come more interesting once JUnit 5 supports parallel threads officially.